File: frontend/src/api/projects.ts
import request from '@/utils/request'

export interface Project {
  id: number
  name: string
  description?: string | null
  owner_id: number
  visibility: string
  status: string
  role?: string | null
  case_count?: number
  element_count?: number
  suite_count?: number
  created_at: string
  updated_at: string
}

export interface ProjectMember {
  membership_id: number | null
  user_id: number
  username?: string | null
  role: string
}

export interface UserCandidate {
  id: number
  username: string
  email: string
}

export interface PageData<T> {
  total: number
  page: number
  page_size: number
  items: T[]
}

export function listProjects(params?: { page?: number; page_size?: number; visibility?: string }) {
  return request.get<PageData<Project>>('/projects', { params })
}

export function createProject(data: { name: string; description?: string; visibility?: string }) {
  return request.post<Project>('/projects', data)
}

export function getProject(id: number) {
  return request.get<Project>(`/projects/${id}`)
}

export function updateProject(id: number, data: Partial<Project>) {
  return request.put<Project>(`/projects/${id}`, data)
}

export function deleteProject(id: number) {
  return request.delete<void>(`/projects/${id}`)
}

export function listMembers(projectId: number) {
  return request.get<ProjectMember[]>(`/projects/${projectId}/members`)
}

export function memberCandidates(projectId: number, keyword: string) {
  return request.get<UserCandidate[]>(`/projects/${projectId}/member-candidates`, {
    params: { keyword, limit: 20 },
  })
}

export function addMember(projectId: number, data: { user_id: number; role: string }) {
  return request.post<ProjectMember>(`/projects/${projectId}/members`, data)
}

export function updateMember(projectId: number, userId: number, role: string) {
  return request.patch<ProjectMember>(`/projects/${projectId}/members/${userId}`, { role })
}

export function removeMember(projectId: number, userId: number) {
  return request.delete<void>(`/projects/${projectId}/members/${userId}`)
}

File: backend/app/api/projects.py
from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user, get_project_permission, require_project_role
from app.core.database import get_db
from app.models import Project, User
from app.schemas.project import (
    PageResult,
    ProjectCreate,
    ProjectMemberCreate,
    ProjectMemberOut,
    ProjectMemberUpdate,
    ProjectOut,
    ProjectUpdate,
    UserCandidateOut,
)
from app.services import project_service
from app.utils.pagination import get_pagination

router = APIRouter(prefix="/projects", tags=["项目管理"])


@router.get("", response_model=PageResult)
async def list_projects(
    pagination=Depends(get_pagination),
    visibility: str = "all",
    scope: str = "",
    role: str = "",
    keyword: str = "",
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await project_service.list_projects(
        db,
        user,
        pagination=pagination,
        visibility=visibility,
        scope=scope,
        role=role,
        keyword=keyword,
    )


@router.post("", response_model=ProjectOut, status_code=status.HTTP_201_CREATED)
async def create_project(
    body: ProjectCreate,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await project_service.create_project(db, body, user.id)


@router.get("/{project_id}", response_model=ProjectOut)
async def get_project(
    project_id: int,
    perm: tuple[Project, str | None] = Depends(get_project_permission),
    db: AsyncSession = Depends(get_db),
):
    project, role = perm
    return await project_service.get_project(db, project, role)


@router.put("/{project_id}", response_model=ProjectOut)
async def update_project(
    project_id: int,
    body: ProjectUpdate,
    perm: tuple[Project, str | None] = Depends(require_project_role("owner", "admin")),
    db: AsyncSession = Depends(get_db),
):
    project, role = perm
    return await project_service.update_project(db, project, body, role)


@router.delete("/{project_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_project(
    project_id: int,
    perm: tuple[Project, str | None] = Depends(require_project_role("owner")),
    db: AsyncSession = Depends(get_db),
):
    project, _role = perm
    await project_service.delete_project(db, project)


@router.get("/{project_id}/members", response_model=list[ProjectMemberOut])
async def list_members(
    project_id: int,
    perm: tuple[Project, str | None] = Depends(get_project_permission),
    db: AsyncSession = Depends(get_db),
):
    project, _role = perm
    return await project_service.list_members(db, project)


@router.get("/{project_id}/member-candidates", response_model=list[UserCandidateOut])
async def member_candidates(
    project_id: int,
    keyword: str = "",
    limit: int = Query(10, ge=1, le=20),
    perm: tuple[Project, str | None] = Depends(require_project_role("owner", "admin")),
    db: AsyncSession = Depends(get_db),
):
    project, _role = perm
    return await project_service.member_candidates(db, project, keyword, limit)


@router.post("/{project_id}/members", response_model=ProjectMemberOut, status_code=status.HTTP_201_CREATED)
async def add_member(
    project_id: int,
    body: ProjectMemberCreate,
    perm: tuple[Project, str | None] = Depends(require_project_role("owner", "admin")),
    db: AsyncSession = Depends(get_db),
):
    project, _role = perm
    return await project_service.add_member(db, project, body)


@router.patch("/{project_id}/members/{user_id}", response_model=ProjectMemberOut)
async def update_member(
    project_id: int,
    user_id: int,
    body: ProjectMemberUpdate,
    perm: tuple[Project, str | None] = Depends(require_project_role("owner", "admin")),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    project, caller_role = perm
    return await project_service.update_member(db, project, user, user_id, body, caller_role)


@router.delete("/{project_id}/members/{user_id}", status_code=status.HTTP_204_NO_CONTENT)
async def remove_member(
    project_id: int,
    user_id: int,
    perm: tuple[Project, str | None] = Depends(require_project_role("owner", "admin")),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    project, caller_role = perm
    await project_service.remove_member(db, project, user, user_id, caller_role)
